docs/source/usage/workflows/add_element.rst

```diff
@@ -0,0 +1,87 @@
+.. _usage-workflows-add-element:
+
+Add New Beamline Elements
+=========================
+
+In ImpactX, one can easily add new beamline elements as a user.
+There are multiple ways to add new elements to ImpactX, you can pick the one that fits your needs best.
+
+
+Python Programmable Element
+---------------------------
+
+Using the :ref:`ImpactX Python interface <usage-python>`, a custom element named :py:class:`impactx.elements.Programmable` can be defined to advance particles using NumPy, CuPy, Numba, PyTorch or any other compatible Python library.
+
+The Programmable element can implement a custom element in two ways:
+
+* Push the whole container, by assigning a ``push`` function or
+* Push the reference particle and beam particles in two individual functions (``beam_particles`` and ``ref_particle``).
+
+Per ImpactX convention, the reference particle is updated *before* the beam particles are pushed.
+
+Detailed examples that show usage of the programmable element are:
+
+* :ref:`FODO cell <examples-fodo-programmable>`: implements a user-defined drift
+* :ref:`15 stage laser-plasma accelerator <examples-ml-surrogate>`: implements a user-defined LPA accelerator element using a neural network surrogate via PyTorch
+
+Detailed particle computing interfaces are presented in the `pyAMReX examples <https://pyamrex.readthedocs.io/en/latest/usage/compute.html#particles>`__.
+
+
+Linear Map
+----------
+
+.. note::
+
+   We plan to add a simple, linear map element that can be configured in user input.
+   Follow `issue #538 <https://github.com/ECP-WarpX/impactx/issues/538>`__ for progress.
+
+
+C++ Element
+-----------
+
```

Comment:
We might want to remark that this workflow is valid for (thin or thick) elements described by explicit maps.  Thick elements with soft-edged fringe fields require additional slicing, to perform integration of the equations for the linear map, before applying this to the particles.  We could say that detailed documentation of implementation of soft-edged elements  is to be addressed later.

Reply:
Good point.

I agree we should elaborate more on how to add thick elements with fringe field in that API. Looking at the current soft-edged elements though: the part we describe here, the push of reference particles and beam particles, is identical, right? I think the detail that differs is what one implements inside the _body_ of the here described APIs, correct?

Reply:
Added a suggestion above, let me know if that sounds reasonable for now?

Reply:
Yes, that's true.  Ok, this looks good.